### PR TITLE
Feat(mailer): add support for dynamic template data

### PR DESCRIPTION
**Pull request by imarquezc**
In this pull request I updated `send_grid_mailer` in order to support dynamic templates data.

This was only possible upgrading sendgrid-ruby gem to `5.3.0`

This is [the PR](https://github.com/sendgrid/sendgrid-ruby/pull/300/files) where the feature was added.

Also, the api v5+ of sendgrid introduced some breaking changes [check the list here](https://github.com/sendgrid/sendgrid-ruby/blob/main/CHANGELOG.md). This PR also fixes those changes.

Finally small updates were made to the tests because of the changes.

### CODE REVIEW

**Review by monkeyci — COMMENTED** [review]
Some files could not be reviewed due to errors:
<details>
<summary>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming</summary>
<pre>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming
.rubocop.yml: Style/MethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/PredicateName has the wrong namespace - should be Naming
.rubocop.yml: Style/VariableName has the wrong namespace - should be Naming
.rubocop.yml: Lint/EndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/DefEndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Style/AccessorMethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/AsciiIdentifiers has the wrong namespace - should be Naming
.rubocop.yml: Style/ClassAndModuleCamelCase has the wrong namespace - should be Naming
.rubocop.yml: Style/ConstantName has the wrong namespace - should be Naming
.rubocop.yml: Style/FlipFlop has the wrong namespace - should be Lint
.rubocop.yml: Lint/BlockAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/ConditionPosition has the wrong namespace - should be Layout
Error: The `Style/TrailingCommaInLiteral` cop no longer exists. Please use `Style/TrailingCommaInArrayLiteral` and/or `Style/TrailingCommaInHashLiteral` instead.
(obsolete configuration found in .rubocop.yml, please update it)
The `Lint/InvalidCharacterLiteral` cop has been removed since it was never being actually triggered.
The `Style/OpMethod` cop has been renamed and moved to `Naming/BinaryOperatorParameterName`.
obsolete parameter EnforcedStyle (for Style/Encoding) found in .rubocop.yml
Style/Encoding no longer supports styles. The "never" behavior is always assumed.
obsolete parameter SupportedStyles (for Style/Encoding) found in .rubocop.yml
obsolete parameter MaxLineLength (for Style/IfUnlessModifier) found in .rubocop.yml
`Style/IfUnlessModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
obsolete parameter MaxLineLength (for Style/WhileUntilModifier) found in .rubocop.yml
`Style/WhileUntilModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
</pre>
</details>

**Review by monkeyci — COMMENTED** [review]
Some files could not be reviewed due to errors:
<details>
<summary>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming</summary>
<pre>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming
.rubocop.yml: Style/MethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/PredicateName has the wrong namespace - should be Naming
.rubocop.yml: Style/VariableName has the wrong namespace - should be Naming
.rubocop.yml: Lint/EndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/DefEndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Style/AccessorMethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/AsciiIdentifiers has the wrong namespace - should be Naming
.rubocop.yml: Style/ClassAndModuleCamelCase has the wrong namespace - should be Naming
.rubocop.yml: Style/ConstantName has the wrong namespace - should be Naming
.rubocop.yml: Style/FlipFlop has the wrong namespace - should be Lint
.rubocop.yml: Lint/BlockAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/ConditionPosition has the wrong namespace - should be Layout
Error: The `Style/TrailingCommaInLiteral` cop no longer exists. Please use `Style/TrailingCommaInArrayLiteral` and/or `Style/TrailingCommaInHashLiteral` instead.
(obsolete configuration found in .rubocop.yml, please update it)
The `Lint/InvalidCharacterLiteral` cop has been removed since it was never being actually triggered.
The `Style/OpMethod` cop has been renamed and moved to `Naming/BinaryOperatorParameterName`.
obsolete parameter EnforcedStyle (for Style/Encoding) found in .rubocop.yml
Style/Encoding no longer supports styles. The "never" behavior is always assumed.
obsolete parameter SupportedStyles (for Style/Encoding) found in .rubocop.yml
obsolete parameter MaxLineLength (for Style/IfUnlessModifier) found in .rubocop.yml
`Style/IfUnlessModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
obsolete parameter MaxLineLength (for Style/WhileUntilModifier) found in .rubocop.yml
`Style/WhileUntilModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
</pre>
</details>

File: spec/dummy/app/assets/config/manifest.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+//= link_tree ../images
+//= link_directory ../javascripts .js
+//= link_directory ../stylesheets .css
```
Comment (monkeyci) [inline spec/dummy/app/assets/config/manifest.js:3]: Expected space or tab after '//' in comment      spaced-comment<br>Newline required at end of file but not found    eol-last

File: spec/dummy/app/assets/config/manifest.js [hunk 2]
```diff
@@ -0,0 +1,3 @@
+//= link_tree ../images
+//= link_directory ../javascripts .js
```
Comment (monkeyci) [inline spec/dummy/app/assets/config/manifest.js:2]: Expected space or tab after '//' in comment      spaced-comment

**Review by monkeyci — COMMENTED** [review]
Some files could not be reviewed due to errors:
<details>
<summary>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming</summary>
<pre>.rubocop.yml: Style/FileName has the wrong namespace - should be Naming
.rubocop.yml: Style/MethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/PredicateName has the wrong namespace - should be Naming
.rubocop.yml: Style/VariableName has the wrong namespace - should be Naming
.rubocop.yml: Lint/EndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/DefEndAlignment has the wrong namespace - should be Layout
.rubocop.yml: Style/AccessorMethodName has the wrong namespace - should be Naming
.rubocop.yml: Style/AsciiIdentifiers has the wrong namespace - should be Naming
.rubocop.yml: Style/ClassAndModuleCamelCase has the wrong namespace - should be Naming
.rubocop.yml: Style/ConstantName has the wrong namespace - should be Naming
.rubocop.yml: Style/FlipFlop has the wrong namespace - should be Lint
.rubocop.yml: Lint/BlockAlignment has the wrong namespace - should be Layout
.rubocop.yml: Lint/ConditionPosition has the wrong namespace - should be Layout
Error: The `Style/TrailingCommaInLiteral` cop no longer exists. Please use `Style/TrailingCommaInArrayLiteral` and/or `Style/TrailingCommaInHashLiteral` instead.
(obsolete configuration found in .rubocop.yml, please update it)
The `Lint/InvalidCharacterLiteral` cop has been removed since it was never being actually triggered.
The `Style/OpMethod` cop has been renamed and moved to `Naming/BinaryOperatorParameterName`.
obsolete parameter EnforcedStyle (for Style/Encoding) found in .rubocop.yml
Style/Encoding no longer supports styles. The "never" behavior is always assumed.
obsolete parameter SupportedStyles (for Style/Encoding) found in .rubocop.yml
obsolete parameter MaxLineLength (for Style/IfUnlessModifier) found in .rubocop.yml
`Style/IfUnlessModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
obsolete parameter MaxLineLength (for Style/WhileUntilModifier) found in .rubocop.yml
`Style/WhileUntilModifier: MaxLineLength` has been removed. Use `Metrics/LineLength: Max` instead
</pre>
</details>

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 3]
```diff
@@ -30,6 +30,30 @@
...
+    end
+
+    it "add dynamic_data to data hash" do
+      subject.dynamic_template_data({ other_key: 'other_value' })
+      expect(personalization.dynamic_template_data.size).to eq(2)
+    end
+
+    it "merges new keys into dynamic_data hash" do
+      subject.dynamic_template_data({ other_key: 'other_value' })
+      expect(personalization.dynamic_template_data).to eq({ key: 'value', other_key: 'other_value' })
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:53]: Redundant curly braces around a hash parameter.<br>Line is too long. [101/100]

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 4]
```diff
@@ -30,6 +30,30 @@
...
+      expect(personalization.dynamic_template_data.size).to eq(1)
+    end
+
+    it "add dynamic_data to data hash" do
+      subject.dynamic_template_data({ other_key: 'other_value' })
+      expect(personalization.dynamic_template_data.size).to eq(2)
+    end
+
+    it "merges new keys into dynamic_data hash" do
+      subject.dynamic_template_data({ other_key: 'other_value' })
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:52]: Name your test subject if you need to reference it explicitly.<br>Redundant curly braces around a hash parameter.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 5]
```diff
@@ -30,6 +30,30 @@
...
+    it "creates substitution with valid data" do
+      expect(@substitution).to eq({ key: 'value' })
+    end
+
+    it "adds substitution to personalization object" do
+      expect(personalization.dynamic_template_data.size).to eq(1)
+    end
+
+    it "add dynamic_data to data hash" do
+      subject.dynamic_template_data({ other_key: 'other_value' })
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:47]: Name your test subject if you need to reference it explicitly.<br>Redundant curly braces around a hash parameter.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 6]
```diff
@@ -30,6 +30,30 @@
     end
   end
 
+  describe "#dynamic_template_data" do
+    before do
+      @substitution = subject.dynamic_template_data({ key: 'value' })
+    end
+
+    it "creates substitution with valid data" do
+      expect(@substitution).to eq({ key: 'value' })
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:39]: Replace instance variable with local variable or let.<br>Redundant curly braces around a hash parameter.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 7]
```diff
@@ -30,6 +30,30 @@
     end
   end
 
+  describe "#dynamic_template_data" do
+    before do
+      @substitution = subject.dynamic_template_data({ key: 'value' })
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:35]: Name your test subject if you need to reference it explicitly.<br>Redundant curly braces around a hash parameter.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 8]
```diff
@@ -17,7 +17,7 @@
     end
 
     it "creates substitution with valid data" do
-      expect(@substitution.substitution).to eq("%subject%" => "Hi!")
+      expect(@substitution).to eq("%subject%" => "Hi!")
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:20]: Replace instance variable with local variable or let.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 9]
```diff
@@ -96,6 +120,14 @@
     end
   end
 
+  describe "#add_category" do
+    it "adds categories to mail object" do
+      subject.add_category("category1")
+      subject.add_category("category2")
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:126]: Name your test subject if you need to reference it explicitly.

File: spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb [hunk 10]
```diff
@@ -96,6 +120,14 @@
     end
   end
 
+  describe "#add_category" do
+    it "adds categories to mail object" do
+      subject.add_category("category1")
```
Comment (monkeyci) [inline spec/dummy/spec/lib/send_grid_mailer/definition_spec.rb:125]: Name your test subject if you need to reference it explicitly.

**Review by difernandez — CHANGES_REQUESTED** [review]
Sorry for the slow review, but thanks for the PR, support for dynamic templates is a very welcome addition 🎉 ! I haven't yet been able to test the feature though, so I'll update my review when I do.

There's just one thing I could find that should be looked into.

#21 adds support for dev environment, after merging that into this branch and fixing the issue I mentioned, I think this would be ready for merging into master!

**Comment by difernandez** [comment]
I tested it, including the related PR #21, and it all seems to be working well! When the other PR and this one are ready, I'll approve and merge 🚀

**Review by difernandez — APPROVED** [review]
The CI problem will be checked at a later date, we'll merge this for now. Thanks again for your contribution!

**Pull request merged**